compiler-rt/lib/tsan/rtl/tsan_platform_linux.cpp

```diff
@@ -267,6 +267,14 @@ static void ReExecIfNeeded() {
   if (reexec) {
     // Don't check the address space since we're going to re-exec anyway.
   } else if (!CheckAndProtect(false, false, false)) {
+    // ASLR personality check.
+    // N.B. 'personality' is sometimes forbidden by sandboxes, so we only call
+    // this as a last resort (when the memory mapping is incompatible and TSan
+    // would fail anyway).
+    int old_personality = personality(0xffffffff);
```

Comment:
We probably could test this by defining own personality function in a test that will simply abort. This way we can have a list of functions that must not be called by the runtime (prohibited by sandboxes).
Though my bet is that it will fail on some platforms for some reasons. So probably the test should be limited to Linux/x86 only.

Reply:
I'm not quite sure how I could add this to the TSan tests, since the sandbox behavior (e.g., allowing/disallowing 'personality') is dependent on the vendor's configuration.

Taken to an extreme, there's probably some sandbox out there that disallows 'printf', but it would be unreasonable to avoid printf entirely in TSan.

Reply:
I meant adding this to a test:
```
extern "C" int personality(unsigned long) { abort(); }
```

Such test should abort w/o this fix.
Probably better to add to sanitizer_common tests so that all sanitizers are tested.

Of course, there can be sandboxes that prohibit too much and sanitizers won't work under these sandboxes. We can't fix that. But we do want to avoid personality and we want it to not regress tomorrow.

Reply:
Ah, I see! I've added `compiler-rt/test/sanitizer_common/TestCases/Linux/sandbox_forbidden_functions.cpp`. I tested that the test failed with TSan in the absence of this fix, and passes with this fix.

Note that it does have a false positive when TSan is run with high-entropy ASLR: in that environment, calling 'personality' (to check/disable ASLR) is the intended behavior.

Reply:
If it would be only about bots, then I would say let's try to add it (perhaps with an additional comment).
But it will also fail on local developer `ninja check-sanitizer` runs, right?

Reply:
It will fail on local developer `ninja check-sanitizer` if they are running with very high entropy ASLR.

Reply:
Then I afraid it's probably better to disable the test.
But also work under that sandbox will probably break soon.

Reply:
I've removed the test in the latest commit (https://github.com/llvm/llvm-project/pull/79334/commits/0b4c8464b5045e80eccd9acc11e7312d6caa3c88).

With the re-exec patch + this patch:
- low entropy ASLR, no sandbox: TSan works
- low entropy ASLR, with sandbox: TSan works
- high entropy ASLR, no sandbox: TSan works
- high entropy ASLR, with sandbox: TSan will trigger a sandbox violation, but TSan didn't work anyway with high entropy ASLR. Moreover, this is currently only a hypothetical case.